File: app/controllers/assessment/LanguageInspectionController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.assessment;

import be.objectify.deadbolt.java.actions.Dynamic;
import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Pattern;
import be.objectify.deadbolt.java.actions.Restrict;
import controllers.base.BaseController;
import impl.mail.EmailComposer;
import io.ebean.DB;
import io.ebean.ExpressionList;
import io.ebean.FetchConfig;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import models.assessment.Comment;
import models.assessment.LanguageInspection;
import models.exam.Exam;
import models.user.User;
import org.apache.pekko.actor.ActorSystem;
import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.data.DynamicForm;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;
import sanitizers.Attrs;
import sanitizers.CommentSanitizer;
import scala.concurrent.duration.Duration;
import security.Authenticated;

public class LanguageInspectionController extends BaseController {

    @Inject
    protected EmailComposer emailComposer;

    @Inject
    protected ActorSystem actor;

    private final Logger logger = LoggerFactory.getLogger(LanguageInspectionController.class);

    @Dynamic(value = "inspector or admin", meta = "pattern=CAN_INSPECT_LANGUAGE,role=ADMIN,anyMatch=true")
    public Result listInspections(Optional<String> month, Optional<Long> start, Optional<Long> end) {
        ExpressionList<LanguageInspection> query = DB.find(LanguageInspection.class)
            .fetch("exam")
            .fetch("exam.course")
            .fetch("exam.creator", "firstName, lastName, email, userIdentifier")
            .fetch("exam.parent.examOwners", "firstName, lastName, email, userIdentifier")
            .fetch("exam.examLanguages", FetchConfig.ofQuery())
            .fetch("statement")
            .fetch("creator", "firstName, lastName, email, userIdentifier")
            .fetch("assignee", "firstName, lastName, email, userIdentifier")
            .where()
            .ne("exam.state", Exam.State.DELETED);

        if (start.isPresent() || end.isPresent()) {
            if (start.isPresent()) {
                DateTime startDate = new DateTime(start.get()).plusDays(1).withTimeAtStartOfDay();
                query = query.ge("finishedAt", startDate.toDate());
            }

            if (end.isPresent()) {
                DateTime endDate = new DateTime(end.get()).plusDays(1).withTimeAtStartOfDay();
                query = query.lt("finishedAt", endDate.toDate());
            }
        } else if (month.isPresent()) {
            String decoded = URLDecoder.decode(month.get(), StandardCharsets.UTF_8);
            DateTime startWithMonth = DateTime.parse(decoded).withMillisOfDay(0);
            DateTime endWithMonth = startWithMonth.plusMonths(1);
            query = query.between("finishedAt", startWithMonth.toDate(), endWithMonth.toDate());
        }

        return ok(query.findSet());
    }

    @Authenticated
    @Restrict({ @Group("ADMIN"), @Group("TEACHER"), @Group("SUPPORT") })
    public Result createInspection(Http.Request request) {
        DynamicForm df = formFactory.form().bindFromRequest(request);
        Long examId = Long.parseLong(df.get("examId"));
        Exam exam = DB.find(Exam.class, examId);
        if (exam == null) {
            return notFound("i18n_error_exam_not_found");
        }
        if (exam.getLanguageInspection() != null) {
            return forbidden("already sent for inspection");
        }
        if (!exam.getSubjectToLanguageInspection()) {
            return forbidden("not allowed to send for inspection");
        }
        LanguageInspection inspection = new LanguageInspection();
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        inspection.setCreatorWithDate(user);
        inspection.setExam(exam);
        inspection.save();
        return ok();
    }

    @Authenticated
    @Pattern(value = "CAN_INSPECT_LANGUAGE")
    public Result assignInspection(Long id, Http.Request request) {
        LanguageInspection inspection = DB.find(LanguageInspection.class, id);
        if (inspection == null) {
            return notFound("Inspection not found");
        }
        if (inspection.getAssignee() != null) {
            return forbidden("Inspection already taken");
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        inspection.setModifierWithDate(user);
        inspection.setAssignee(user);
        inspection.setStartedAt(new Date());
        inspection.update();
        return ok();
    }

    private Optional<Result> checkInspection(LanguageInspection inspection) {
        if (inspection.getStartedAt() == null) {
            return Optional.of(forbidden("Inspection not assigned"));
        }
        if (inspection.getFinishedAt() != null) {
            return Optional.of(forbidden("Inspection already finalized"));
        }
        return Optional.empty();
    }

    @Authenticated
    @Pattern(value = "CAN_INSPECT_LANGUAGE")
    public Result setApproval(Long id, Http.Request request) {
        DynamicForm df = formFactory.form().bindFromRequest(request);
        boolean isApproved = Boolean.parseBoolean(df.get("approved"));
        LanguageInspection inspection = DB.find(LanguageInspection.class, id);
        if (inspection == null) {
            return notFound("Inspection not found");
        }
        return checkInspection(inspection).orElseGet(() -> {
                if (inspection.getStatement() == null || inspection.getStatement().getComment().isEmpty()) {
                    return forbidden("No statement given");
                }
                inspection.setFinishedAt(new Date());
                inspection.setApproved(isApproved);

                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                inspection.setModifierWithDate(user);
                inspection.update();

                Set<User> recipients = inspection.getExam().getParent().getExamOwners();
                actor
                    .scheduler()
                    .scheduleOnce(
                        Duration.create(1, TimeUnit.SECONDS),
                        () -> {
                            for (User recipient : recipients) {
                                emailComposer.composeLanguageInspectionFinishedMessage(recipient, user, inspection);
                                logger.info("Language inspection finalization email sent to {}", recipient.getEmail());
                            }
                        },
                        actor.dispatcher()
                    );

                return ok();
            });
    }

    @Authenticated
    @With(CommentSanitizer.class)
    @Pattern(value = "CAN_INSPECT_LANGUAGE")
    public Result setStatement(Long id, Http.Request request) {
        Optional<String> text = request.attrs().getOptional(Attrs.COMMENT);
        if (text.isEmpty()) {
            return badRequest();
        }
        LanguageInspection inspection = DB.find(LanguageInspection.class, id);
        if (inspection == null) {
            return notFound("Inspection not found");
        }
        return checkInspection(inspection).orElseGet(() -> {
                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                Comment statement = inspection.getStatement();
                if (statement == null) {
                    statement = new Comment();
                    statement.setCreatorWithDate(user);
                    statement.save();
                    inspection.setStatement(statement);
                    inspection.update();
                }
                statement.setComment(text.get());
                statement.setModifierWithDate(user);
                statement.update();
                inspection.setModifierWithDate(user);
                inspection.update();
                return ok(inspection);
            });
    }
}

File: app/controllers/assessment/LanguageInspectionController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.assessment

import controllers.base.scala.ExamBaseController
import impl.mail.EmailComposer
import io.ebean.{DB, ExpressionList, FetchConfig}
import miscellaneous.scala.DbApiHelper
import models.assessment.{Comment, LanguageInspection}
import models.exam.Exam
import models.user.Permission.Type
import models.user.Role
import org.apache.pekko.actor.ActorSystem
import org.joda.time.DateTime
import play.api.Logging
import play.api.libs.json.JsValue
import play.api.mvc.*
import security.scala.Auth.{AuthenticatedAction, authorized}
import security.scala.{Auth, AuthExecutionContext, CombinedRoleAndPermissionFilter, PermissionFilter}
import system.AuditedAction
import validation.scala.CommentValidator
import validation.scala.core.{ScalaAttrs, Validators}

import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Date
import javax.inject.Inject
import scala.concurrent.duration.DurationInt
import scala.jdk.CollectionConverters.CollectionHasAsScala

class LanguageInspectionController @Inject() (
    val controllerComponents: ControllerComponents,
    val validators: Validators,
    val authenticated: AuthenticatedAction,
    val audited: AuditedAction,
    val actorSystem: ActorSystem,
    val emailComposer: EmailComposer,
    implicit val ec: AuthExecutionContext
) extends BaseController
    with ExamBaseController
    with DbApiHelper
    with Logging:

  def listInspections(month: Option[String], start: Option[Long], end: Option[Long]): Action[AnyContent] =
    authenticated.andThen(CombinedRoleAndPermissionFilter.anyMatch(Type.CAN_INSPECT_LANGUAGE, Role.Name.ADMIN)) {
      request =>
        val query = DB
          .find(classOf[LanguageInspection])
          .fetch("exam")
          .fetch("exam.course")
          .fetch("exam.creator", "firstName, lastName, email, userIdentifier")
          .fetch("exam.parent.examOwners", "firstName, lastName, email, userIdentifier")
          .fetch("exam.examLanguages", FetchConfig.ofQuery)
          .fetch("statement")
          .fetch("creator", "firstName, lastName, email, userIdentifier")
          .fetch("assignee", "firstName, lastName, email, userIdentifier")
          .where
          .ne("exam.state", Exam.State.DELETED)
        val filteredQuery = applyDateFilters(query, month, start, end)
        Ok(filteredQuery.distinct.asJson)
    }

  def createInspection(): Action[JsValue] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)).andThen(audited))(parse.json) {
        request =>
          val examId = (request.body \ "examId").as[Long]
          Option(DB.find(classOf[Exam], examId)) match
            case Some(exam) =>
              if Option(exam.getLanguageInspection).isDefined then Forbidden("already sent for inspection")
              else if !exam.getSubjectToLanguageInspection then Forbidden("not allowed for language inspection")
              else
                val inspection = new LanguageInspection
                inspection.setExam(exam)
                inspection.setCreatorWithDate(request.attrs(Auth.ATTR_USER))
                inspection.save()
                Ok
            case None => BadRequest
      }

  def assignInspection(id: Long): Action[AnyContent] =
    authenticated.andThen(PermissionFilter(Type.CAN_INSPECT_LANGUAGE)) { request =>
      Option(DB.find(classOf[LanguageInspection], id)) match
        case Some(inspection) =>
          if Option(inspection.getAssignee).isDefined then Forbidden("Inspection already assigned")
          val user = request.attrs(Auth.ATTR_USER)
          inspection.setModifierWithDate(user)
          inspection.setAssignee(user)
          inspection.setStartedAt(new Date())
          inspection.update()
          Ok
        case None => NotFound("inspection not found")
    }

  def setApproval(id: Long): Action[JsValue] =
    authenticated.andThen(PermissionFilter(Type.CAN_INSPECT_LANGUAGE))(parse.json).andThen(audited) { request =>
      (request.body \ "approved").asOpt[Boolean] match
        case Some(approval) =>
          Option(DB.find(classOf[LanguageInspection], id)) match
            case Some(inspection) =>
              if Option(inspection.getStartedAt).isEmpty then Forbidden("Inspection not assigned")
              if Option(inspection.getFinishedAt).isEmpty then Forbidden("Inspection already finalized")
              if Option(inspection.getStatement).isEmpty || inspection.getStatement.getComment.isEmpty then
                Forbidden("No statement given")
              inspection.setFinishedAt(new Date)
              inspection.setApproved(approval)
              val user = request.attrs(Auth.ATTR_USER)
              inspection.setModifierWithDate(user)
              inspection.update()

              val recipients = inspection.getExam.getParent.getExamOwners.asScala
              actorSystem.scheduler.scheduleOnce(1.seconds)(recipients.foreach(r =>
                emailComposer.composeLanguageInspectionFinishedMessage(r, user, inspection)
                logger.info(s"Language inspection finalization email sent to ${r.getEmail}")
              ))
              Ok
            case None => NotFound("inspection not found")
        case None => BadRequest
    }

  def setStatement(id: Long): Action[AnyContent] = authenticated
    .andThen(PermissionFilter(Type.CAN_INSPECT_LANGUAGE))
    .andThen(validators.validated(CommentValidator))
    .andThen(audited) { request =>
      request.attrs.get(ScalaAttrs.COMMENT) match
        case Some(comment) =>
          Option(DB.find(classOf[LanguageInspection], id)) match
            case Some(inspection) =>
              if Option(inspection.getStartedAt).isEmpty then Forbidden("Inspection not assigned")
              if Option(inspection.getFinishedAt).isEmpty then Forbidden("Inspection already finalized")
              val user = request.attrs(Auth.ATTR_USER)
              val statement = Option(inspection.getStatement).getOrElse {
                val newComment = new Comment
                newComment.setCreatorWithDate(user)
                newComment.save()
                inspection.setStatement(newComment)
                inspection.update()
                newComment
              }
              statement.setComment(comment)
              statement.setModifierWithDate(user)
              statement.update()
              inspection.setModifierWithDate(user)
              inspection.update()
              Ok(inspection.asJson)
            case None => NotFound("inspection not found")
        case None => BadRequest("No comment given")
    }

  private def applyDateFilters(
      query: ExpressionList[LanguageInspection],
      month: Option[String],
      start: Option[Long],
      end: Option[Long]
  ): ExpressionList[LanguageInspection] =
    (start, end, month) match
      // Case 1: Range filter (start and/or end)
      case (Some(_), _, _) | (_, Some(_), _) =>
        val withStart = start.fold(query) { s =>
          val startDate = new DateTime(s).plusDays(1).withTimeAtStartOfDay()
          query.ge("finishedAt", startDate.toDate)
        }
        end.fold(withStart) { e =>
          val endDate = new DateTime(e).plusDays(1).withTimeAtStartOfDay()
          withStart.lt("finishedAt", endDate.toDate)
        }

      // Case 2: Month filter
      case (None, None, Some(monthStr)) =>
        val decoded      = URLDecoder.decode(monthStr, StandardCharsets.UTF_8)
        val startOfMonth = DateTime.parse(decoded).withMillisOfDay(0)
        val endOfMonth   = startOfMonth.plusMonths(1)
        query.between("finishedAt", startOfMonth.toDate, endOfMonth.toDate)

      // Case 3: No filter
      case _ => query
